File: AdjacencyMatrix.js
// Graph Representation in JavaScript (Adjacency Matrix)

class Graph {
  // Initialize the graph with a specific number of vertices
  constructor(numVertices) {
    this.numVertices = numVertices;

    // Create a V x V matrix and initialize all entries to 0
    this.adjMatrix = this.initializeMatrix(numVertices);
  }

  // Helper function to create and initialize the 2D array
  initializeMatrix(numVertices) {
    let matrix = [];

    for (let i = 0; i < numVertices; i++) {
      matrix[i] = []; // create a new inner array

      for (let j = 0; j < numVertices; j++) {
        matrix[i][j] = 0; // set initial value to 0 (no edge)
      }
    }

    return matrix;
  }

  // Add an edge between vertex u and v
  addEdge(u, v, weight = 1, directed = false) {
    if (
      u >= 0 &&
      u < this.numVertices &&
      v >= 0 &&
      v < this.numVertices
    ) {
      // set the value for the edge from u to v
      this.adjMatrix[u][v] = weight;

      // If undirected, add the reverse edge (v → u)
      if (!directed) {
        this.adjMatrix[v][u] = weight;
      }
    } else {
      console.log("Error: Vertex out of range");
    }
  }

  // Print the adjacency matrix
  printMatrix() {
    for (let i = 0; i < this.numVertices; i++) {
      console.log(this.adjMatrix[i].join(" "));
    }
  }
}

// Create an undirected graph with 4 vertices (0, 1, 2, 3)
const g = new Graph(4);

// Add edges (0,1), (0,2), (1,2), (1,3)
g.addEdge(0, 1);
g.addEdge(0, 2);
g.addEdge(1, 2);
g.addEdge(1, 3);

console.log("Adjacency Matrix:");
g.printMatrix();

File: AdjacencyMatrix.py
import numpy as np

def create_adjacency_matrix_numpy(num_vertices, edges, directed=False):
    # Initialize a V x V numpy array with zeros
    adj_matrix = np.zeros((num_vertices, num_vertices), dtype=int)

    # Populate the matrix with edges
    for u, v in edges:
        if 0 <= u < num_vertices and 0 <= v < num_vertices:
            adj_matrix[u][v] = 1
            if not directed:
                adj_matrix[v][u] = 1
        else:
            print(f"Edge ({u}, {v}) out of range.")
            return None

    return adj_matrix


# Example usage:
num_v = 4
graph_edges = [(0, 1), (0, 2), (1, 2), (1, 3)]

adj_mat_np = create_adjacency_matrix_numpy(num_v, graph_edges)

if adj_mat_np is not None:
    print("Adjacency matrix (NumPy):")
    print(adj_mat_np)
